chain/chain/src/chain.rs

```diff
@@ -1968,6 +1970,160 @@ impl Chain {
+                Ok(Some(hash))
+            } else {
+                let cur_tree_size = (index + 1) * counter;
+                if cur_tree_size > tree_size {
+                    let left_hash =
+                        self.get_node(index * 2, level - 1, counter / 2, tree_size, tree_nodes)?;
+                    let right_hash = self.reconstruct_node(
+                        index * 2 + 1,
+                        level - 1,
+                        counter / 2,
+                        tree_size,
+                        tree_nodes,
+                    )?;
+                    return Ok(Self::combine_maybe_hashes(left_hash, right_hash));
+                }
+                let hash = *self
+                    .mut_store()
+                    .get_block_merkle_tree_from_ordinal(cur_tree_size)?
+                    .get_path()
+                    .last()
+                    .ok_or_else(|| ErrorKind::Other("Merkle tree node missing".to_string()))?;
+                tree_nodes.insert((index, level), hash);
+                Ok(Some(hash))
+            }
+        }
+    }
+
+    /// Reconstruct node at given position (index, level). If the node does not exist, return `None`.
+    fn reconstruct_node(
+        &mut self,
+        index: u64,
+        level: u64,
+        counter: u64,
+        tree_size: u64,
+        tree_nodes: &mut HashMap<(u64, u64), MerkleHash>,
+    ) -> Result<Option<MerkleHash>, Error> {
+        if let Some(hash) = tree_nodes.get(&(index, level)) {
+            Ok(Some(*hash))
+        } else {
+            if level == 0 {
+                if index >= tree_size {
+                    return Ok(None);
+                }
+                let hash = *self.mut_store().get_block_hash_from_ordinal(index)?;
+                tree_nodes.insert((index, level), hash);
+                Ok(Some(hash))
+            } else {
+                let left_hash =
+                    self.get_node(index * 2, level - 1, counter / 2, tree_size, tree_nodes)?;
+                let right_hash = self.reconstruct_node(
+                    index * 2 + 1,
+                    level - 1,
+                    counter / 2,
+                    tree_size,
+                    tree_nodes,
+                )?;
+                let hash = Self::combine_maybe_hashes(left_hash, right_hash);
+                if let Some(h) = hash {
+                    tree_nodes.insert((index, level), h);
+                }
+
+                Ok(hash)
+            }
+        }
+    }
+
+    /// Get merkle proof for block with hash `block_hash` in the merkle tree of `head_block_hash`.
+    pub fn get_block_proof(
+        &mut self,
+        block_hash: &CryptoHash,
+        head_block_hash: &CryptoHash,
+    ) -> Result<MerklePath, Error> {
+        let leaf_index = self.mut_store().get_block_merkle_tree(block_hash)?.size();
+        let tree_size = self.mut_store().get_block_merkle_tree(head_block_hash)?.size();
+        if leaf_index >= tree_size {
```

Comment:
> We should be able to have proofs that check against the head itself.

For a given block, the merkle tree of blocks include every block up to the previous block (otherwise we cannot compute the hash of the current block). Therefore, in the merkle tree of a given block there is no way to prove itself :)

Reply:
We should specialcase and return an empty proof for example?
It's just weird to not be able to send proofs to a light client that can be checked against the block they have :)